### PR TITLE
Bluetooth: Mesh: Move Replay Protect to seperate module

### PR DESCRIPTION
**Pull request by LingaoM**
Move RPL to seperate module, and remove it in `bt_mesh`
structure.

~~Fixes proxy configuration not check by replay.~~
~~Fixes secure beacon iv update not store new RPL.~~

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

### CODE REVIEW

File: subsys/bluetooth/mesh/proxy.c [hunk 1]
```diff
@@ -291,6 +292,12 @@ static void proxy_cfg(struct bt_mesh_proxy_client *client)
 		return;
 	}
 
+	if (bt_mesh_replay_check(&rx, NULL)) {
+		BT_WARN("Replay: src 0x%04x dst 0x%04x seq 0x%06x",
+			rx.ctx.addr, rx.ctx.recv_dst, rx.seq);
+		return;
+	}
+
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/proxy.c:300]: This is a bug. Could you split this out into a separate PR, so we can add it to 2.4?

File: subsys/bluetooth/mesh/rpl.c [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+#define LOG_MODULE_NAME bt_mesh_rpl
+#include "common/log.h"
+
+#include "mesh.h"
+#include "adv.h"
+#include "net.h"
+#include "rpl.h"
+#include "settings.h"
+
+struct bt_mesh_rpl rpl_tabel[CONFIG_BT_MESH_CRPL];
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/rpl.c:33]:
```suggestion
struct bt_mesh_rpl rpl_table[CONFIG_BT_MESH_CRPL];
```

Or just `rpl` or `replay_list`? replay_list_table is a bit redundant

File: subsys/bluetooth/mesh/rpl.c [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+	for (i = 0; i < ARRAY_SIZE(rpl_tabel); i++) {
+		struct bt_mesh_rpl *rpl = &rpl_tabel[i];
+
+		if (rpl->src) {
+			if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+				bt_mesh_clear_rpl((void *)rpl);
+			}
+		}
+	}
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/rpl.c:120]: Also a potential bug?

File: subsys/bluetooth/mesh/rpl.h [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	uint32_t seq;
+};
+
+void bt_mesh_rpl_reset(void);
+bool bt_mesh_replay_check(struct bt_mesh_net_rx *rx,
+			struct bt_mesh_rpl **match);
+void bt_mesh_rpl_clear(void);
+void bt_mesh_update_rpl(struct bt_mesh_rpl *rpl,
+			struct bt_mesh_net_rx *rx);
+struct bt_mesh_rpl *bt_mesh_rpl_alloc(void);
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/rpl.h:22]: Looks like you inherited a mix of terminology and function name formats here. Could you perhaps align all functions to be `bt_mesh_rpl_<...>`, for consistency?

File: subsys/bluetooth/mesh/settings.c [hunk 5]
```diff
@@ -2474,6 +2355,48 @@ int bt_mesh_model_data_store(struct bt_mesh_model *mod, bool vnd,
 	return err;
 }
 
+void bt_mesh_store_rpl(void *entry)
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/settings.c:2358]: Why `void *`?

File: subsys/bluetooth/mesh/settings.c [hunk 6]
```diff
@@ -2474,6 +2355,48 @@ int bt_mesh_model_data_store(struct bt_mesh_model *mod, bool vnd,
...
+
+	snprintk(path, sizeof(path), "bt/mesh/RPL/%x", rpl->src);
+
+	err = settings_save_one(path, &val, sizeof(val));
+	if (err) {
+		BT_ERR("Failed to store RPL %s value", log_strdup(path));
+	} else {
+		BT_DBG("Stored RPL %s value", log_strdup(path));
+	}
+}
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/settings.c:2379]: This change removes the staggered storing of replay entries, which may reduce the flash lifetime significantly. I think it would be better to not change the settings module for this patch, and expose the `rpl_tabel` as an extern in the rpl header, until we get around to fixing #19850. After this, we can retain the old functionality without exposing the internal data structure at all.

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

**Comment by LingaoM** [comment]
Remove bug label,due to #28356 solved these issues. wait for PR(#28356) the merger, and then solve @trond-snekvik modification proposal.

File: subsys/bluetooth/mesh/proxy.c [hunk 7]
```diff
@@ -300,6 +301,12 @@ static void proxy_cfg(struct bt_mesh_proxy_client *client)
 		return;
 	}
 
+	if (bt_mesh_rpl_check(&rx, NULL)) {
+		BT_WARN("Replay: src 0x%04x dst 0x%04x seq 0x%06x",
+			rx.ctx.addr, rx.ctx.recv_dst, rx.seq);
+		return;
+	}
+
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/proxy.c:309]: This is now a duplicate check, and should be removed.

File: subsys/bluetooth/mesh/rpl.h [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+struct bt_mesh_rpl {
+	uint16_t src;
+	bool  old_iv;
+#if defined(CONFIG_BT_SETTINGS)
+	bool  store;
+#endif
+	uint32_t seq;
+};
+
+extern struct bt_mesh_rpl replay_list[CONFIG_BT_MESH_CRPL];
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/rpl.h:19]: `replay_list` should also have the `bt_mesh_` prefix like the functions in this header

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

**Review by trond-snekvik — APPROVED** [review]

File: subsys/bluetooth/mesh/rpl.h [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+struct bt_mesh_rpl {
+	uint16_t src;
+	bool  old_iv;
+#if defined(CONFIG_BT_SETTINGS)
+	bool  store;
+#endif
+	uint32_t seq;
+};
+
+extern struct bt_mesh_rpl bt_mesh_replay_list[CONFIG_BT_MESH_CRPL];
```
Comment (jhedberg) [inline subsys/bluetooth/mesh/rpl.h:19]: Why not call the variable `bt_mesh_rpl`? Also, it comes across as a bit odd that you expose these various accessor functions for the RPL but then anyway expose the full thing, basically making the encapsulation unenforceable.

File: subsys/bluetooth/mesh/settings.c [hunk 10]
```diff
@@ -284,9 +285,9 @@ static struct bt_mesh_rpl *rpl_find(uint16_t src)
 {
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(bt_mesh.rpl); i++) {
-		if (bt_mesh.rpl[i].src == src) {
-			return &bt_mesh.rpl[i];
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_replay_list); i++) {
+		if (bt_mesh_replay_list[i].src == src) {
+			return &bt_mesh_replay_list[i];
```
Comment (jhedberg) [inline subsys/bluetooth/mesh/settings.c:290]: Seems like this `rpl_find()` and `rpl_alloc()` further below are the reasons why you have to expose the entire struct  outside of the new `rpl.c`? Wouldn't it make sense to move these to `rpl.c` as well and then keep the struct definition and variable internal to `rpl.c`?

**Review by jhedberg — CHANGES_REQUESTED** [review]

**Review by jhedberg — APPROVED** [review]
